File: frontend/src/search/fuzzy.ts
// A tiny, dependency-free fuzzy matcher. Higher scores are better; null means
// "no match" so the caller can drop the candidate. It rewards contiguous
// substring hits over scattered subsequences, and word-boundary/prefix matches
// over mid-word ones, which is what makes "nap" find "network.admin.port" while
// still ranking "admin" above it. Kept small on purpose - at our scale a linear
// scan of tens of thousands of short strings is well under a frame, so there is
// no need for an index or a third-party library.

// A boundary is anything that is not a letter or digit ("." "_" "/" "-" space),
// i.e. the start of a new word in an identifier or path.
function isBoundary(ch: string | undefined): boolean {
  return ch === undefined || !/[a-z0-9]/i.test(ch);
}

/** Score how well `query` matches `text`. Returns null when it does not. */
export function fuzzyScore(query: string, text: string): number | null {
  if (!query) return 0;
  const q = query.toLowerCase();
  const t = text.toLowerCase();

  // Contiguous substring: the strongest signal.
  const sub = t.indexOf(q);
  if (sub !== -1) {
    let s = 900 - sub; // earlier is better
    if (isBoundary(t[sub - 1])) s += 120; // starts a word (or the string)
    if (t === q) s += 400; // exact
    return s - (t.length - q.length) * 0.5; // tighter is better
  }

  // Subsequence: every query char appears in order.
  let ti = 0;
  let score = 0;
  let prev = -2;
  for (let qi = 0; qi < q.length; qi++) {
    const c = q[qi];
    let hit = -1;
    while (ti < t.length) {
      if (t[ti] === c) {
        hit = ti;
        break;
      }
      ti++;
    }
    if (hit === -1) return null;
    score += 10;
    if (hit === prev + 1) score += 8; // adjacent to the previous match
    if (isBoundary(t[hit - 1])) score += 6; // on a word boundary
    prev = hit;
    ti = hit + 1;
  }
  return score - t.length * 0.1; // gently prefer shorter fields
}

/** Best score of `query` across several weighted fields (null if none match). */
export function fuzzyBest(
  query: string,
  fields: (string | undefined)[],
  weights?: number[],
): number | null {
  let best: number | null = null;
  for (let i = 0; i < fields.length; i++) {
    const f = fields[i];
    if (!f) continue;
    const raw = fuzzyScore(query, f);
    if (raw === null) continue;
    const val = raw * (weights?.[i] ?? 1);
    if (best === null || val > best) best = val;
  }
  return best;
}

File: backend/internal/search/fuzzy.go
package search

import "strings"

// fuzzyScore mirrors frontend/src/search/fuzzy.ts so the in-memory candidate
// scan behaves the same on both sides. Higher is better; ok=false means no
// match. It rewards contiguous substrings over scattered subsequences and
// word-boundary/prefix matches over mid-word ones. Identifiers here are ASCII
// (parameter names, instance names), so byte indexing is sufficient.
func fuzzyScore(query, text string) (float64, bool) {
	if query == "" {
		return 0, true
	}
	q := strings.ToLower(query)
	t := strings.ToLower(text)

	if idx := strings.Index(t, q); idx != -1 {
		s := 900.0 - float64(idx)
		if isBoundary(t, idx-1) {
			s += 120
		}
		if t == q {
			s += 400
		}
		return s - float64(len(t)-len(q))*0.5, true
	}

	ti := 0
	score := 0.0
	prev := -2
	for i := 0; i < len(q); i++ {
		c := q[i]
		hit := -1
		for ti < len(t) {
			if t[ti] == c {
				hit = ti
				break
			}
			ti++
		}
		if hit == -1 {
			return 0, false
		}
		score += 10
		if hit == prev+1 {
			score += 8
		}
		if isBoundary(t, hit-1) {
			score += 6
		}
		prev = hit
		ti = hit + 1
	}
	return score - float64(len(t))*0.1, true
}

// isBoundary reports whether the character before position i starts a new word
// (the start of the string, or any non-alphanumeric separator).
func isBoundary(t string, i int) bool {
	if i < 0 {
		return true
	}
	ch := t[i]
	if ch >= 'a' && ch <= 'z' {
		return false
	}
	if ch >= '0' && ch <= '9' {
		return false
	}
	return true
}

// bestScore is the max fuzzy score of a query across the title (weighted) and
// the keyword haystack; ok=false when neither matches.
func bestScore(query, title, keywords string) (float64, bool) {
	best := 0.0
	found := false
	if s, ok := fuzzyScore(query, title); ok {
		best = s * 2
		found = true
	}
	if s, ok := fuzzyScore(query, keywords); ok && s > best {
		best = s
		found = true
	}
	return best, found
}
